wgpu-hal/src/dx12/shader_compilation.rs

```diff
@@ -0,0 +1,260 @@
+        stage_bit: wgt::ShaderStages,
+        full_stage: String,
+        dxc_container: &DxcContainer,
+    ) -> (
+        Result<crate::dx12::CompiledShader, crate::PipelineError>,
+        log::Level,
+    ) {
+        let mut compile_flags = arrayvec::ArrayVec::<&str, 3>::new_const();
+        compile_flags.push("-Ges"); // d3dcompiler::D3DCOMPILE_ENABLE_STRICTNESS
+        if device
+            .private_caps
+            .instance_flags
+            .contains(crate::InstanceFlags::DEBUG)
+        {
+            compile_flags.push("-Zi"); /* d3dcompiler::D3DCOMPILE_SKIP_OPTIMIZATION */
+            compile_flags.push("-Od"); /* d3dcompiler::D3DCOMPILE_DEBUG */
+        }
+
+        let blob = match dxc_container
+            .library
+            .create_blob_with_encoding_from_str(source)
+            .map_err(|e| crate::PipelineError::Linkage(stage_bit, format!("DXC blob error: {}", e)))
+        {
+            Ok(blob) => blob,
+            Err(e) => return (Err(e), log::Level::Error),
+        };
+
+        let compiled = dxc_container.compiler.compile(
+            &blob,
+            source_name,
+            raw_ep,
+            &full_stage,
+            &compile_flags,
+            None, // TODO: idk what the include handler does
+            &[],
+        );
+
+        let (result, log_level) = match compiled {
+            Ok(dxc_result) => match dxc_result.get_result() {
+                Ok(dxc_blob) => (
+                    Ok(crate::dx12::CompiledShader::Dxc(dxc_blob.to_vec())),
+                    log::Level::Info,
+                ),
+                Err(e) => (
+                    Err(crate::PipelineError::Linkage(
+                        stage_bit,
+                        format!("DXC compile error: {}", e),
+                    )),
+                    log::Level::Error,
+                ),
+            },
+            Err(e) => (
+                Err(crate::PipelineError::Linkage(
+                    stage_bit,
+                    format!("DXC compile error: {:?}", e),
+                )),
+                log::Level::Error,
+            ),
+        };
+
+        (result, log_level)
+    }
+
+    impl From<hassle_rs::HassleError> for crate::DeviceError {
+        fn from(value: hassle_rs::HassleError) -> Self {
+            match value {
+                hassle_rs::HassleError::Win32Error(e) => {
+                    // TODO: This returns an HRESULT, should we try and use the associated Windows error message?
+                    log::error!("Win32 error: {e:?}");
+                    crate::DeviceError::Lost
+                }
+                hassle_rs::HassleError::LoadLibraryError { filename, inner } => {
+                    log::error!("Failed to load dxc library {filename:?}. Inner error: {inner:?}");
+                    crate::DeviceError::Lost
+                }
+                hassle_rs::HassleError::LibLoadingError(e) => {
+                    log::error!("Failed to load dxc library. {e:?}");
```

Comment:
My additional text on these errors is probably redundant, as I'm also displaying the hassle error and whatever their text for it is.